File: Python/libraries/datatypes-timex-expression/datatypes_timex_expression/timex_format.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from .timex_constants import Constants
from .timex_helpers import TimexHelpers
from .timex_date_helpers import TimexDateHelpers

class TimexFormat:

    @staticmethod
    def format(timex):

        types = timex.types

        if Constants.TIMEX_TYPES_PRESENT in types:
            return 'PRESENT_REF'

        if ((Constants.TIMEX_TYPES_DATETIMERANGE in types or
            Constants.TIMEX_TYPES_DATERANGE in types or
            Constants.TIMEX_TYPES_TIMERANGE in types) and
            Constants.TIMEX_TYPES_DURATION in types):
            timex_range = TimexHelpers.expand_datetime_range(timex)
            return '({},{},{})'.format(
                TimexFormat.format(timex_range.start),
                TimexFormat.format(timex_range.end),
                TimexFormat.format(timex_range.duration))

        if Constants.TIMEX_TYPES_DATETIMERANGE in types:
            return '{}{}'.format(
                TimexFormat.format_date(timex),
                TimexFormat.format_time_range(timex))

        if Constants.TIMEX_TYPES_DATERANGE in types:
            return TimexFormat.format_date_range(timex)

        if Constants.TIMEX_TYPES_TIMERANGE in types:
            return TimexFormat.format_time_range(timex)
        
        if Constants.TIMEX_TYPES_DATETIME in types:
            return '{}{}'.format(
                TimexFormat.format_date(timex),
                TimexFormat.format_time(timex))
        
        if Constants.TIMEX_TYPES_DURATION in types:
            return TimexFormat.format_duration(timex)

        if Constants.TIMEX_TYPES_DATE in types:
            return TimexFormat.format_date(timex)

        if Constants.TIMEX_TYPES_TIME in types:
            return TimexFormat.format_time(timex)

        return ''

    @staticmethod
    def format_duration(timex):
        if timex.years:
            return 'P{}Y'.format(timex.years)

        if timex.months:
            return 'P{}M'.format(timex.months)

        if timex.weeks:
            return 'P{}W'.format(timex.weeks)

        if timex.days:
            return 'P{}D'.format(timex.days)

        if timex.hours:
            return 'PT{}H'.format(timex.hours)

        if timex.minutes:
            return 'PT{}M'.format(timex.minutes)

        if timex.seconds:
            return 'PT{}S'.format(timex.seconds)

        return ''

    @staticmethod
    def format_time(timex):
        if timex.minute == 0 and timex.second == 0:
            return 'T{}'.format(TimexDateHelpers.fixed_format_number(timex.hour, 2))

        if timex.second == 0:
            return 'T{}:{}'.format(
                TimexDateHelpers.fixed_format_number(timex.hour, 2),
                TimexDateHelpers.fixed_format_number(timex.minute, 2))

        return 'T{}:{}:{}'.format(
            TimexDateHelpers.fixed_format_number(timex.hour, 2),
            TimexDateHelpers.fixed_format_number(timex.minute, 2),
            TimexDateHelpers.fixed_format_number(timex.second, 2))

    @staticmethod
    def format_date(timex):
        if timex.year is not None and timex.month is not None and timex.day_of_month is not None:
            return '{}-{}-{}'.format(
                TimexDateHelpers.fixed_format_number(timex.Year, 4),
                TimexDateHelpers.fixed_format_number(timex.Month, 2),
                TimexDateHelpers.fixed_format_number(timex.DayOfMonth, 2))

        if timex.month is not None and timex.day_of_month is not None:
            return 'XXXX-{}-{}'.format(
                TimexDateHelpers.fixed_format_number(timex.month, 2),
                TimexDateHelpers.fixed_format_number(timex.day_of_month, 2))

        if timex.day_of_week is not None:
            return 'XXXX-WXX-{}'.format(timex.day_of_week)

        return ''

    @staticmethod
    def format_date_range(timex):
        if timex.year is not None and timex.week_of_year is not None and timex.weekend is not None:
            return '{}-W{}-WE'.format(TimexDateHelpers.fixed_format_number(timex.year, 4), TimexDateHelpers.fixed_format_number(timex.week_of_year, 2))

        if timex.year is not None and timex.week_of_year is not None:
            return '{}-W{}'.format(TimexDateHelpers.fixed_format_number(timex.year, 4), TimexDateHelpers.fixed_format_number(timex.week_of_year, 2))

        if timex.year is not None and timex.season is not None:
            return '{}-{}'.format(TimexDateHelpers.fixed_format_number(timex.year, 4), timex.season)

        if timex.season:
            return '{}'.format(timex.season)

        if timex.year is not None and timex.month is not None:
            return '{}-{}'.format(TimexDateHelpers.fixed_format_number(timex.year, 4), TimexDateHelpers.fixed_format_number(timex.Month, 2))

        if timex.year:
            return '{}'.format(TimexDateHelpers.fixed_format_number(timex.year, 4))

        if timex.month is not None and timex.week_of_month is not None and timex.day_of_week is not None:
            return 'XXXX-{}-WXX-{}-{}'.format(TimexDateHelpers.fixed_format_number(timex.month, 2), timex.week_of_month, timex.day_of_week)

        if timex.month is not None and timex.week_of_month is not None:
            return 'XXXX-{}-WXX-{}'.format(TimexDateHelpers.fixed_format_number(timex.month, 2), timex.week_of_month)

        if timex.month:
            return 'XXXX-{}'.format(TimexDateHelpers.fixed_format_number(timex.month, 2))

        return ''

    @staticmethod
    def format_time_range(timex):
        if timex.part_of_day != None:
            return 'T{}'.format(timex.part_of_day)

        return ''

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Globalization;

namespace Microsoft.Recognizers.Text.DataTypes.TimexExpression
{
    public static class TimexFormat
    {
        public static string Format(TimexProperty timex)
        {
            if (timex.IsPresent)
            {
                return "PRESENT_REF";
            }

            if (timex.IsDuration && (timex.IsDateTimeRange || timex.IsDateRange || timex.IsTimeRange))
            {
                var range = TimexHelpers.ExpandDateTimeRange(timex);
                return $"({Format(range.Start)},{Format(range.End)},{Format(range.Duration)})";
            }

            if (timex.IsDateTimeRange)
            {
                return $"{FormatDate(timex)}{FormatTimeRange(timex)}";
            }

            if (timex.IsDateRange)
            {
                return $"{FormatDateRange(timex)}";
            }

            if (timex.IsTimeRange)
            {
                return $"{FormatTimeRange(timex)}";
            }

            if (timex.IsDateTime)
            {
                return $"{FormatDate(timex)}{FormatTime(timex)}";
            }

            if (timex.IsDuration)
            {
                return $"{FormatDuration(timex)}";
            }

            if (timex.IsDate)
            {
                return $"{FormatDate(timex)}";
            }

            if (timex.IsTime)
            {
                return $"{FormatTime(timex)}";
            }

            return string.Empty;
        }

        private static string FormatDuration(TimexProperty timex)
        {
            var timexList = new List<string> { };
            if (timex.Years != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Year, timex.Years ?? Constants.InvalidValue));
            }

            if (timex.Months != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Month, timex.Months ?? Constants.InvalidValue));
            }

            if (timex.Weeks != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Week, timex.Weeks ?? Constants.InvalidValue));
            }

            if (timex.Days != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Day, timex.Days ?? Constants.InvalidValue));
            }

            if (timex.Hours != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Hour, timex.Hours ?? Constants.InvalidValue));
            }

            if (timex.Minutes != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Minute, timex.Minutes ?? Constants.InvalidValue));
            }

            if (timex.Seconds != null)
            {
                timexList.Add(TimexHelpers.GenerateDurationTimex(TimexUnit.Second, timex.Seconds ?? Constants.InvalidValue));
            }

            return TimexHelpers.GenerateCompoundDurationTimex(timexList);
        }

        private static string FormatTime(TimexProperty timex)
        {
            if (timex.Minute == 0 && timex.Second == 0)
            {
                return $"T{TimexDateHelpers.FixedFormatNumber(timex.Hour, 2)}";
            }

            if (timex.Second == 0)
            {
                return $"T{TimexDateHelpers.FixedFormatNumber(timex.Hour, 2)}:{TimexDateHelpers.FixedFormatNumber(timex.Minute, 2)}";
            }

            return $"T{TimexDateHelpers.FixedFormatNumber(timex.Hour, 2)}:{TimexDateHelpers.FixedFormatNumber(timex.Minute, 2)}:{TimexDateHelpers.FixedFormatNumber(timex.Second, 2)}";
        }

        private static string FormatDate(TimexProperty timex)
        {
            return TimexHelpers.GenerateDateTimex(timex.Year ?? Constants.InvalidValue, timex.WeekOfYear ?? (timex.Month ?? Constants.InvalidValue), timex.DayOfWeek != null ? timex.DayOfWeek.Value : timex.DayOfMonth ?? Constants.InvalidValue, timex.WeekOfMonth ?? Constants.InvalidValue, timex.DayOfWeek != null);
        }

        private static string FormatDateRange(TimexProperty timex)
        {
            if (timex.Year != null && timex.WeekOfYear != null && timex.Weekend != null)
            {
                return $"{TimexDateHelpers.FixedFormatNumber(timex.Year, 4)}-W{TimexDateHelpers.FixedFormatNumber(timex.WeekOfYear, 2)}-WE";
            }

            if (timex.Year != null && timex.WeekOfYear != null)
            {
                return $"{TimexDateHelpers.FixedFormatNumber(timex.Year, 4)}-W{TimexDateHelpers.FixedFormatNumber(timex.WeekOfYear, 2)}";
            }

            if (timex.Year != null && timex.Month != null && timex.WeekOfMonth != null)
            {
                return $"{TimexDateHelpers.FixedFormatNumber(timex.Year, 4)}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}-W{TimexDateHelpers.FixedFormatNumber(timex.WeekOfMonth, 2)}";
            }

            if (timex.Year != null && timex.Season != null)
            {
                return $"{TimexDateHelpers.FixedFormatNumber(timex.Year, 4)}-{timex.Season}";
            }

            if (timex.Season != null)
            {
                return $"{timex.Season}";
            }

            if (timex.Year != null && timex.Month != null)
            {
                return $"{TimexDateHelpers.FixedFormatNumber(timex.Year, 4)}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}";
            }

            if (timex.Year != null)
            {
                return $"{TimexDateHelpers.FixedFormatNumber(timex.Year, 4)}";
            }

            if (timex.Month != null && timex.WeekOfMonth != null && timex.DayOfWeek != null)
            {
                return $"{Constants.TimexFuzzyYear}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}-{Constants.TimexFuzzyWeek}-{timex.WeekOfMonth}-{timex.DayOfWeek}";
            }

            if (timex.Month != null && timex.WeekOfMonth != null)
            {
                return $"{Constants.TimexFuzzyYear}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}-W{timex.WeekOfMonth?.ToString("D2", CultureInfo.InvariantCulture)}";
            }

            if (timex.Month != null)
            {
                return $"{Constants.TimexFuzzyYear}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}";
            }

            return string.Empty;
        }

        private static string FormatTimeRange(TimexProperty timex)
        {
            if (timex.PartOfDay != null)
            {
                return $"T{timex.PartOfDay}";
            }

            return string.Empty;
        }
    }
}
